### ISSUE TITLE
Docs remain after deleting Buddydpress Group

### ISSUE THREAD

**Issue by reachingmike** [boonebgorges/buddypress-docs#579]
When deleting a Buddypress group that has linked docs the docs remain in the database.
Shouldn't they be deleted or at least provide a method to achieve this?
Is it likely to cause any issues if I were to delete these manually from the database?

**Comment by boonebgorges** [comment]
@reachingmike Hello, and thanks for the ticket!

I can't imagine any problem you'd have by deleting the documents manually.

Deleting the Docs automatically seems problematic to me - there are plenty of cases where you may want to keep the Docs even though the associated group is gone (especially if the Docs are not private). Perhaps I could add an opt-in prompt to the group delete confirmation screen. We should also improve the way that the Dashboard tools work, so that it's possible to filter based on certain group associations. This would make it easier for admins to do manual cleanup later on.

**Comment by reachingmike** [comment]
Thanks for the quick reply @boonebgorges.

I had considered the auto deletion scenario wouldn't suit everyone but, an opt-in would be really helpful for my scenario - or maybe a 'delete-all' button somewhere perhaps.

**Comment by dcavins** [comment]
I'm also wary of auto-deleting docs when the parent container is deleted. However, we do follow that pattern with folders, I believe, so maybe it's OK in this case, too. I'd rather have the group-related docs revert to author-access-only docs (and I think we do something similar somewhere on an action), but if a doc is put in the trash, the author can still access it. So maybe my doubts are overblown.

**Comment by boonebgorges** [comment]
I like the idea of proposing a cleanup routine as part of the group deletion process. It looks like we can hook to `bp_after_group_delete_admin` and add some markup. I guess we'd give three options: a. Send associated docs to the trash, b. Change doc permissions, c. Do nothing. For (b), we could either give more fine-grained options (admin can select from "author-access-only" or "public"), or maybe we preemptively decide that docs from a public group should become public, otherwise they should be made author-only.

As I think about it, it seems like offering this option becomes potentially frustrating if there's no clear way to find the orphaned Docs after the group has been deleted. This would be a good opportunity to improve some filtering mechanisms in the Dashboard, but there's a problem: once a group is deleted, there's no obvious way for us to provide a user-friendly filter. (It's just be the `bp_docs_associated_with_123` taxonomy term, where 123 stands for a group that no longer exists in the database.) I don't know how to deal with this.

**Comment by reachingmike** [comment]
Could you offer any advise on what would be a reliable sql query I could use to delete docs for a group manually?

**Comment by boonebgorges** [comment]
I wouldn't suggest deleting purely in SQL, as you're likely to miss stuff. But you could do something like this:

```
$atts = get_posts( array(
    'post_type' => 'attachment',
    'post_status' => 'inherit',
    'post_parent' => $deleted_doc_id,
    'fields' => 'ids',
    'posts_per_page' => -1,
) );

foreach ( $atts as $att ) {
    wp_delete_post( $att );
}
```

Untested, but something along those lines is probably your best bet.

**Comment by reachingmike** [comment]
Hi, I'm struggling with your example to use php to delete docs for a group - Could you please explain what is $deleted_doc_id for me? It's not the group id is it?

**Comment by boonebgorges** [comment]
No. You'd have to have some additional logic. This is totally untested and contains some pseudo-code, but it will look something like:

```
add_action( 'groups_before_delete_group', function( $group_id ) {
	$q = new BP_Docs_Query( array(
		'group_id' => $group_id,
		'posts_per_page' => -1,
	) );

	$query_args = $q->get_wp_query();

	// You'll probably need to remove hidden protection to do this programatically.
	remove_action( 'pre_get_posts', 'bp_docs_general_access_protection', 28 );
	
	$doc_query = new WP_Query( $query_args );
	foreach ( $doc_query->posts as $post ) {
		// delete doc attachments
		$atts = get_posts( array(
			'post_type' => 'attachment',
			'post_status' => 'inherit',
			'post_parent' => $post->ID,
			'fields' => 'ids',
			'posts_per_page' => -1,
		) );

		foreach ( $atts as $att ) {
			wp_delete_post( $att );
		}

		// delete the doc
		wp_delete_post( $post->ID );
	}
	
	add_action( 'pre_get_posts', 'bp_docs_general_access_protection', 28 );
} );
```
